[WIP] Added scott and silvermann rules for bandwidth estimation.

icfly2 opened the pull request:
This addresses  #10467, Bandwidth estimation for kernel density estimates.

String defining the bandwidth selection type is overwritten with fit.

This is a work in progress.

The cross-validation method to select the bandwidth should be added too.

Commit:
Added scott and silvermann rules for bandwidth estimation. String is overwritten with fit

jnothman commented on sklearn/neighbors/kde.py:
```diff
@@ -123,6 +127,11 @@ def fit(self, X, y=None):
         X = check_array(X, order='C', dtype=DTYPE)
+        
+        if self.bandwidth == 'scott':
+            self.bandwidth = 3.5*X.std()/len(X)**0.3333
```
Instead of overwriting the parameter, please introduce an estimated attribute `bandwidth_`.

Numpy's histogramming function implements this with higher precision constants ((24.0 * np.pi**0.5) ** 1/3. instead of 3.5; 1/3. instead of 0.3333). Do we care? Should we use higher precision constants here?

Also, I'd prefer spacing like `3.5 * X.std() / len(X)**0.3333` which emphasises the order of operations, while the spaces add readability.

icfly2 replied:
Just to clarify, `bandwidth_` should contain the string defining the method right? The `bandwidth` attribute is used as a float later in other functions and I'd rather not change those functions too.

I'll change it to the higher precision constants for consistency, in practice this will have little effect.

tagomatech replied:
Can I ask why you would even use `3.5` instead of `3.49` as in [Scott's paper](http://www.stat.cmu.edu/~rnugent/PCMI2016/papers/ScottBandwidth.pdf) (5. DATA-BASED HISTOGRAMS, p.608)?
By the same token, it is true Python is not very good at high-precision calculations but why don't you just let python deal itself with the `1./3.` fraction instead of plugging `0.333` or `0.3333` or etc... ?
Apologies if I'm totally missing the point.

jnothman reviewed (commented):
Please add tests. Is there a reason to choose scott and silverman? Knowing little about these functions, how do these relate to the options available for bins in np.histogram?

jnothman commented on sklearn/neighbors/kde.py:
```diff
@@ -86,9 +87,12 @@ def __init__(self, bandwidth=1.0, algorithm='auto',
-
-        if bandwidth <= 0:
+        
+        if bandwidth in ['scott','silvermann']:
```
Space after comma, please

jnothman reviewed (changes requested):
Please also add parameter documentation.

Commit:
Added updated scott and silvermann rule, also added cv method

jnothman commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-60832b40bd4ca68dc636302e5e456516c5015d1a)

**new alerts:**

* 1 for Syntax error

---

*Comment posted by [lgtm.com](https://lgtm.com)*

Commit:
working cv, now needs tests and some tuning

Commit:
Added tests, but failing

Commit:
tests in a passing

icfly2 commented:
I've added tests to the bits I added, though they only check if the program runs, there is no validation of the quality of the result, which is somewhat dependent on the random sampling used. I've used RandomState(0) to initialise the test so I guess if would be feasible to fix the outcome and test against it, though that would mean any improvement would fail the test and would need changing. 

I don't really understand the output of Travis: 
- It ask for `bandwidth_` to be set in `__init__` which it is.
- It says `rvs`  got a `random_state`  variable which is in the documentation: [link](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.rvs.html#scipy.stats.rv_continuous.rvs)
- The failing ranking of the cv results is actually the result of another bug #10529 I just filed. 

I also have no idea to interpret the Appveyor output.

jnothman commented on sklearn/neighbors/kde.py:
```diff
@@ -120,9 +131,61 @@ def fit(self, X, y=None):
+                                                metric_params = self.metric_params),
+                                    {'bandwidth': bandwidth_range},
+                                    cv=20
+                                    )
```
refit=False

jnothman commented on sklearn/neighbors/kde.py:
```diff
@@ -86,9 +92,14 @@ def __init__(self, bandwidth=1.0, algorithm='auto',
         # so we can't do this kind of logic in __init__
         self._choose_algorithm(self.algorithm, self.metric)
+        
+        if bandwidth in ['scott', 'silvermann','cv']:
```
Please set self.bandwidth in `__init__` with no validation. In fit, estimate and set `bandwidth_`, which you should use in score_samples

jnothman commented on sklearn/neighbors/kde.py:
```diff
@@ -120,9 +131,61 @@ def fit(self, X, y=None):
+        elif self.bandwidth_ == 'silvermann':
+            self.bandwidth = (X.shape[0] * (X.shape[1] + 2) / 4.)**(-1. / (X.shape[1] + 4))
+        elif self.bandwidth_ == 'cv':
+            steps = 10
```
I suspect this predetermines too many constants. But certainly it wastes computation rebuilding the tree in each underlying fit. It may also to be me conventional to provide KernelDensityCV to do this...

cmarmo commented:
I'm closing this pull request as the related issue has been closed by #22993.
